Invalid dependency / injection resolution for producers

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:0.21.1:build (default) on project apicurio-registry-storage-kafka: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: Found 2 deployment problems: 
[ERROR] [1] Ambiguous dependencies for type java.util.function.Function<org.apache.kafka.clients.producer.ProducerRecord<io.apicurio.registry.kafka.proto.Reg$UUID, io.apicurio.registry.kafka.proto.Reg$RegistryValue>, java.util.concurrent.CompletableFuture<org.apache.kafka.clients.producer.RecordMetadata>> and qualifiers [@Default]
[ERROR] 	- java member: io.apicurio.registry.kafka.KafkaRegistryStorage#registryProducer
[ERROR] 	- declared on CLASS bean [types=[io.apicurio.registry.kafka.KafkaRegistryStorageHandle, io.apicurio.registry.storage.RegistryStorage, io.apicurio.registry.storage.impl.SimpleMapRegistryStorage, java.lang.Object, io.apicurio.registry.kafka.KafkaRegistryStorage, io.apicurio.registry.storage.impl.AbstractMapRegistryStorage], qualifiers=[@Default, @Any], target=io.apicurio.registry.kafka.KafkaRegistryStorage]
[ERROR] 	- available beans:
[ERROR] 		- PRODUCER METHOD bean [types=[java.lang.AutoCloseable, java.util.function.Function<org.apache.kafka.clients.producer.ProducerRecord<K, V>, java.util.concurrent.CompletableFuture<org.apache.kafka.clients.producer.RecordMetadata>>, io.apicurio.registry.kafka.utils.ProducerActions<io.apicurio.registry.kafka.proto.Reg$UUID, io.apicurio.registry.kafka.proto.Reg$RegistryValue>, java.lang.Object], qualifiers=[@Default, @Any], target=io.apicurio.registry.kafka.utils.ProducerActions<io.apicurio.registry.kafka.proto.Reg$UUID, io.apicurio.registry.kafka.proto.Reg$RegistryValue> registryProducer(java.util.Properties properties), declaringBean=io.apicurio.registry.kafka.KafkaRegistryConfiguration]
[ERROR] 		- PRODUCER METHOD bean [types=[java.lang.AutoCloseable, java.util.function.Function<org.apache.kafka.clients.producer.ProducerRecord<K, V>, java.util.concurrent.CompletableFuture<org.apache.kafka.clients.producer.RecordMetadata>>, java.lang.Object, io.apicurio.registry.kafka.utils.ProducerActions<java.lang.Long, io.apicurio.registry.kafka.snapshot.StorageSnapshot>], qualifiers=[@Default, @Any], target=io.apicurio.registry.kafka.utils.ProducerActions<java.lang.Long, io.apicurio.registry.kafka.snapshot.StorageSnapshot> snapshotProducer(java.util.Properties properties), declaringBean=io.apicurio.registry.kafka.KafkaRegistryConfiguration]
[ERROR] [2] Ambiguous dependencies for type java.util.function.Function<org.apache.kafka.clients.producer.ProducerRecord<java.lang.Long, io.apicurio.registry.kafka.snapshot.StorageSnapshot>, java.util.concurrent.CompletableFuture<org.apache.kafka.clients.producer.RecordMetadata>> and qualifiers [@Default]
[ERROR] 	- java member: io.apicurio.registry.kafka.KafkaRegistryStorage#snapshotProducer
[ERROR] 	- declared on CLASS bean [types=[io.apicurio.registry.kafka.KafkaRegistryStorageHandle, io.apicurio.registry.storage.RegistryStorage, io.apicurio.registry.storage.impl.SimpleMapRegistryStorage, java.lang.Object, io.apicurio.registry.kafka.KafkaRegistryStorage, io.apicurio.registry.storage.impl.AbstractMapRegistryStorage], qualifiers=[@Default, @Any], target=io.apicurio.registry.kafka.KafkaRegistryStorage]
[ERROR] 	- available beans:
[ERROR] 		- PRODUCER METHOD bean [types=[java.lang.AutoCloseable, java.util.function.Function<org.apache.kafka.clients.producer.ProducerRecord<K, V>, java.util.concurrent.CompletableFuture<org.apache.kafka.clients.producer.RecordMetadata>>, io.apicurio.registry.kafka.utils.ProducerActions<io.apicurio.registry.kafka.proto.Reg$UUID, io.apicurio.registry.kafka.proto.Reg$RegistryValue>, java.lang.Object], qualifiers=[@Default, @Any], target=io.apicurio.registry.kafka.utils.ProducerActions<io.apicurio.registry.kafka.proto.Reg$UUID, io.apicurio.registry.kafka.proto.Reg$RegistryValue> registryProducer(java.util.Properties properties), declaringBean=io.apicurio.registry.kafka.KafkaRegistryConfiguration]
[ERROR] 		- PRODUCER METHOD bean [types=[java.lang.AutoCloseable, java.util.function.Function<org.apache.kafka.clients.producer.ProducerRecord<K, V>, java.util.concurrent.CompletableFuture<org.apache.kafka.clients.producer.RecordMetadata>>, java.lang.Object, io.apicurio.registry.kafka.utils.ProducerActions<java.lang.Long, io.apicurio.registry.kafka.snapshot.StorageSnapshot>], qualifiers=[@Default, @Any], target=io.apicurio.registry.kafka.utils.ProducerActions<java.lang.Long, io.apicurio.registry.kafka.snapshot.StorageSnapshot> snapshotProducer(java.util.Properties properties), declaringBean=io.apicurio.registry.kafka.KafkaRegistryConfiguration]
```

```java
    @Produces
    @ApplicationScoped
    public ProducerActions<Long, StorageSnapshot> snapshotProducer(
        @KafkaProperties("registry.kafka.snapshot-producer.") Properties properties
    ) {
        return new AsyncProducer<>(
            properties,
            Serdes.Long().serializer(),
            new StorageSnapshotSerde()
        );
    }

    @Produces
    @ApplicationScoped
    public ProducerActions<Reg.UUID, Reg.RegistryValue> registryProducer(
        @KafkaProperties("registry.kafka.registry-producer.") Properties properties
    ) {
        return new AsyncProducer<>(
            properties,
            ProtoSerde.parsedWith(Reg.UUID.parser()),
            ProtoSerde.parsedWith(Reg.RegistryValue.parser())
        );
    }

    @Inject
    Function<ProducerRecord<Reg.UUID, Reg.RegistryValue>, CompletableFuture<RecordMetadata>> registryProducer;

    @Inject
    Function<ProducerRecord<Long, StorageSnapshot>, CompletableFuture<RecordMetadata>> snapshotProducer;


public interface ProducerActions<K, V> extends Function<ProducerRecord<K, V>, CompletableFuture<RecordMetadata>>, AutoCloseable {
}

public class AsyncProducer<K, V> implements ProducerActions<K, V> {

```

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -209,7 +209,6 @@ final class Types {
             // Canonical ParameterizedType with unresolved type variables
             Type[] typeParams = new Type[typeParameters.size()];
             for (int i = 0; i < typeParameters.size(); i++) {
-                // TODO more complicated hierarchy
                 typeParams[i] = resolvedTypeParameters.get(typeParameters.get(i));
             }
             if (resolvedTypeVariablesConsumer != null) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -21,6 +21,7 @@ import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.FieldInfo;
+import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.PrimitiveType.Primitive;
@@ -149,7 +150,7 @@ final class Types {
             } else if (Kind.PARAMETERIZED_TYPE.equals(returnType.kind())) {
                 types = getTypeClosure(returnTypeClassInfo,
                         buildResolvedMap(returnType.asParameterizedType().arguments(), returnTypeClassInfo.typeParameters(),
-                                Collections.emptyMap()),
+                                Collections.emptyMap(), beanDeployment.getIndex()),
                         beanDeployment, null);
             } else {
                 throw new IllegalArgumentException("Unsupported return type");
@@ -175,7 +176,7 @@ final class Types {
             } else if (Kind.PARAMETERIZED_TYPE.equals(fieldType.kind())) {
                 types = getTypeClosure(fieldClassInfo,
                         buildResolvedMap(fieldType.asParameterizedType().arguments(), fieldClassInfo.typeParameters(),
-                                Collections.emptyMap()),
+                                Collections.emptyMap(), beanDeployment.getIndex()),
                         beanDeployment, null);
             } else {
                 throw new IllegalArgumentException("Unsupported return type");
@@ -191,7 +192,7 @@ final class Types {
             types = getTypeClosure(classInfo, Collections.emptyMap(), beanDeployment, null);
         } else {
             types = getTypeClosure(classInfo, buildResolvedMap(typeParameters, typeParameters,
-                    Collections.emptyMap()), beanDeployment, null);
+                    Collections.emptyMap(), beanDeployment.getIndex()), beanDeployment, null);
         }
         return restrictBeanTypes(types, beanDeployment.getAnnotations(classInfo));
     }
@@ -208,6 +209,7 @@ final class Types {
             // Canonical ParameterizedType with unresolved type variables
             Type[] typeParams = new Type[typeParameters.size()];
             for (int i = 0; i < typeParameters.size(); i++) {
+                // TODO more complicated hierarchy
                 typeParams[i] = resolvedTypeParameters.get(typeParameters.get(i));
             }
             if (resolvedTypeVariablesConsumer != null) {
@@ -226,7 +228,7 @@ final class Types {
                 Map<TypeVariable, Type> resolved = Collections.emptyMap();
                 if (Kind.PARAMETERIZED_TYPE.equals(interfaceType.kind())) {
                     resolved = buildResolvedMap(interfaceType.asParameterizedType().arguments(),
-                            interfaceClassInfo.typeParameters(), resolvedTypeParameters);
+                            interfaceClassInfo.typeParameters(), resolvedTypeParameters, beanDeployment.getIndex());
                 }
                 types.addAll(getTypeClosure(interfaceClassInfo, resolved, beanDeployment, resolvedTypeVariablesConsumer));
             }
@@ -239,7 +241,7 @@ final class Types {
                 if (Kind.PARAMETERIZED_TYPE.equals(classInfo.superClassType().kind())) {
                     resolved = buildResolvedMap(classInfo.superClassType().asParameterizedType().arguments(),
                             superClassInfo.typeParameters(),
-                            resolvedTypeParameters);
+                            resolvedTypeParameters, beanDeployment.getIndex());
                 }
                 types.addAll(getTypeClosure(superClassInfo, resolved, beanDeployment, resolvedTypeVariablesConsumer));
             }
@@ -277,18 +279,35 @@ final class Types {
         return types;
     }
 
-    static <T extends Type> Map<TypeVariable, Type> buildResolvedMap(List<T> resolvedTypeVariables,
+    static <T extends Type> Map<TypeVariable, Type> buildResolvedMap(List<T> resolvedArguments,
             List<TypeVariable> typeVariables,
-            Map<TypeVariable, Type> resolvedTypeParameters) {
+            Map<TypeVariable, Type> resolvedTypeParameters, IndexView index) {
         Map<TypeVariable, Type> resolvedMap = new HashMap<>();
-        for (int i = 0; i < resolvedTypeVariables.size(); i++) {
-            Type resolvedTypeVariable = resolvedTypeVariables.get(i);
-            Type resolvedTypeParam = Kind.TYPE_VARIABLE.equals(resolvedTypeVariable.kind())
-                    ? resolvedTypeParameters.getOrDefault(resolvedTypeVariable, resolvedTypeVariable)
-                    : resolvedTypeVariable;
-            resolvedMap.put(typeVariables.get(i), resolvedTypeParam);
+        for (int i = 0; i < resolvedArguments.size(); i++) {
+            resolvedMap.put(typeVariables.get(i), resolveTypeParam(resolvedArguments.get(i), resolvedTypeParameters, index));
         }
         return resolvedMap;
+    }
+
+    static Type resolveTypeParam(Type typeParam, Map<TypeVariable, Type> resolvedTypeParameters, IndexView index) {
+        if (typeParam.kind() == Kind.TYPE_VARIABLE) {
+            return resolvedTypeParameters.getOrDefault(typeParam, typeParam);
+        } else if (typeParam.kind() == Kind.PARAMETERIZED_TYPE) {
+            ParameterizedType parameterizedType = typeParam.asParameterizedType();
+            ClassInfo classInfo = index.getClassByName(parameterizedType.name());
+            if (classInfo != null) {
+                List<TypeVariable> typeParameters = classInfo.typeParameters();
+                List<Type> arguments = parameterizedType.arguments();
+                Map<TypeVariable, Type> resolvedMap = buildResolvedMap(arguments, typeParameters,
+                        resolvedTypeParameters, index);
+                Type[] typeParams = new Type[typeParameters.size()];
+                for (int i = 0; i < typeParameters.size(); i++) {
+                    typeParams[i] = resolveTypeParam(arguments.get(0), resolvedMap, index);
+                }
+                return ParameterizedType.create(parameterizedType.name(), typeParams, null);
+            }
+        }
+        return typeParam;
     }
 
     static String getPackageName(String className) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -302,7 +302,7 @@ final class Types {
                         resolvedTypeParameters, index);
                 Type[] typeParams = new Type[typeParameters.size()];
                 for (int i = 0; i < typeParameters.size(); i++) {
-                    typeParams[i] = resolveTypeParam(arguments.get(0), resolvedMap, index);
+                    typeParams[i] = resolveTypeParam(arguments.get(i), resolvedMap, index);
                 }
                 return ParameterizedType.create(parameterizedType.name(), typeParams, null);
             }

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/generic/GenericProducerHierarchyTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/generic/GenericProducerHierarchyTest.java
@@ -1,0 +1,55 @@
+package io.quarkus.arc.test.producer.generic;
+
+import static org.junit.Assert.assertEquals;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.util.TypeLiteral;
+import javax.inject.Singleton;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class GenericProducerHierarchyTest {
+
+    @Rule
+    public ArcTestContainer container = new ArcTestContainer(Producer.class);
+
+    @SuppressWarnings("serial")
+    @Test
+    public void testPrimitiveProducers() {
+        InstanceHandle<Function<Optional<String>, String>> strHandle = Arc.container()
+                .instance(new TypeLiteral<Function<Optional<String>, String>>() {
+                });
+        Function<Optional<String>, String> producedStr = strHandle.get();
+        assertEquals("foo", producedStr.apply(Optional.of("FOO")));
+
+        InstanceHandle<Function<Optional<Long>, Long>> longHandle = Arc.container()
+                .instance(new TypeLiteral<Function<Optional<Long>, Long>>() {
+                });
+        Function<Optional<Long>, Long> producedLong = longHandle.get();
+        assertEquals(Long.valueOf(-10), producedLong.apply(Optional.of(10l)));
+    }
+
+    @Singleton
+    static class Producer {
+
+        @Produces
+        public Produced<String, String> produce() {
+            return s -> s.get().toLowerCase();
+        }
+
+        @Produces
+        public Produced<Long, Long> produceLong() {
+            return v -> -v.get();
+        }
+    }
+
+    static interface Produced<T, R> extends Function<Optional<T>, R> {
+
+    }
+
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/generic/GenericProducerHierarchyTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/generic/GenericProducerHierarchyTest.java
@@ -3,12 +3,11 @@ package io.quarkus.arc.test.producer.generic;
 import static org.junit.Assert.assertEquals;
 
 import io.quarkus.arc.Arc;
-import io.quarkus.arc.InstanceHandle;
 import io.quarkus.arc.test.ArcTestContainer;
 import java.util.Optional;
 import java.util.function.Function;
 import javax.enterprise.inject.Produces;
-import javax.enterprise.util.TypeLiteral;
+import javax.inject.Inject;
 import javax.inject.Singleton;
 import org.junit.Rule;
 import org.junit.Test;
@@ -16,39 +15,41 @@ import org.junit.Test;
 public class GenericProducerHierarchyTest {
 
     @Rule
-    public ArcTestContainer container = new ArcTestContainer(Producer.class);
+    public ArcTestContainer container = new ArcTestContainer(Producer.class, Registry.class);
 
-    @SuppressWarnings("serial")
     @Test
-    public void testPrimitiveProducers() {
-        InstanceHandle<Function<Optional<String>, String>> strHandle = Arc.container()
-                .instance(new TypeLiteral<Function<Optional<String>, String>>() {
-                });
-        Function<Optional<String>, String> producedStr = strHandle.get();
-        assertEquals("foo", producedStr.apply(Optional.of("FOO")));
-
-        InstanceHandle<Function<Optional<Long>, Long>> longHandle = Arc.container()
-                .instance(new TypeLiteral<Function<Optional<Long>, Long>>() {
-                });
-        Function<Optional<Long>, Long> producedLong = longHandle.get();
-        assertEquals(Long.valueOf(-10), producedLong.apply(Optional.of(10l)));
+    public void testGenericProducers() {
+        Registry registry = Arc.container().instance(Registry.class).get();
+        assertEquals("foo", registry.strProducer.apply(s -> s.toLowerCase()).get());
+        assertEquals(Long.valueOf(-10), registry.longProducer.apply(l -> Long.valueOf(-l)).get());
     }
 
     @Singleton
     static class Producer {
 
         @Produces
-        public Produced<String, String> produce() {
-            return s -> s.get().toLowerCase();
+        public Produced<String, CharSequence> produce() {
+            return f -> Optional.of(f.apply("FOO"));
         }
 
         @Produces
-        public Produced<Long, Long> produceLong() {
-            return v -> -v.get();
+        public Produced<Integer, Long> produceLong() {
+            return f -> Optional.of(f.apply(10));
         }
     }
 
-    static interface Produced<T, R> extends Function<Optional<T>, R> {
+    static interface Produced<T, R> extends Function<Function<T, R>, Optional<R>> {
+
+    }
+
+    @Singleton
+    static class Registry {
+
+        @Inject
+        Function<Function<String, CharSequence>, Optional<CharSequence>> strProducer;
+
+        @Inject
+        Function<Function<Integer, Long>, Optional<Long>> longProducer;
 
     }
 

